psi4/driver/procrouting/response/scf_response.py

```diff
@@ -255,165 +261,398 @@ def _print_output(complete_dict, output):
+           symmetry. When asking for ``states = [3, 3, 3, 3], triplets =
+           "ALSO"`` states (C2v symmetry), ``[2, 2, 2, 2]`` will be of singlet
+           spin symmetry and ``[1, 1, 1, 1]``` will be of triplet spin
+           symmetry.
+    tda :  bool, optional.
+       Should the solver use the Tamm-Dancoff approximation (TDA) or the
+       random-phase approximation (RPA)?
+       Default is ``False``: use RPA.
+       Note that TDA is equivalent to CIS for HF references.
+    r_convergence : float, optional.
+       The convergence threshold for the norm of the residual vector.
+       Default: 1.0e-4
+       Using a tighter convergence threshold here requires tighter SCF ground
+       state convergence threshold. As a rule of thumb, with the SCF ground
+       state density converged to :math:`10^{-N}` (``D_CONVERGENGE = 1.0e-N``),
+       you can afford converging a corresponding TDSCF calculation to
+       :math:`10^{-(N-2)}`.
+       The default value is consistent with the default value for
+       ``D_CONVERGENCE``.
+    maxiter : int, optional
+       Maximum number of iterations.
+       Default: 60
+    guess : str, optional.
+       How should the starting trial vectors be generated?
+       Default: `DENOMINATORS`, i.e. use orbital energy differences to generate
+       guess vectors.
+    verbose : int, optional.
+       How verbose should the solver be?
+       Default: 1
+
+    Notes
+    -----
+    The algorithm employed to solve the non-Hermitian eigenvalue problem (``tda = False``)
+    will fail when the SCF wavefunction has a triplet instability.
+
+    This function can be used for:
+      - restricted singlets: RPA or TDA, any functional
+      - restricted triplets: RPA or TDA, Hartree-Fock only
+      - unresctricted: RPA or TDA, Hartre-Fock and LDA only
+
+    Tighter convergence thresholds will require a larger iterative subspace.
+    The maximum size of the iterative subspace is calculated based on `r_convergence`:
+
+       max_vecs_per_root = -np.log10(r_convergence) * 50
+
+    for the default converegence threshold this gives 200 trial vectors per root and a maximum subspace size
+    of:
+
+       max_ss_size = max_vecs_per_root * n
+
+    where `n` are the number of roots to seek in the given irrep.
+    For each irrep, the algorithm will store up to `max_ss_size` trial vectors
+    before collapsing (restarting) the iterations from the `n` best
+    approximations.
+    """
 
-    #TODO: oscillator strengths
+    # validate input parameters
+    _validate_tdscf(wfn=wfn, states=states, triplets=triplets, guess=guess)
 
-    #TODO: check/handle convergence failures
+    restricted = wfn.same_a_b_orbs()
+
+    # determine how many states per irrep to seek and apportion them between singlets/triplets and irreps.
+    singlets_per_irrep = []
+    triplets_per_irrep = []
+    if isinstance(states, list):
+        if triplets == "ONLY":
```

Comment:
`triplets` has been sanity-checked for case? rhs is going to be lowercase if it passes through `energy(..., **kwargs)`.